js/modules/k6/ws/ws.go

```diff
@@ -436,7 +436,7 @@ func (s *Socket) closeConnection(code int) error {
 			websocket.FormatCloseMessage(code, ""),
 			time.Now().Add(writeWait),
 		)
-		if err != nil {
+		if err != nil && err != websocket.ErrCloseSent {
```

Comment:
Why, though ? I mean it's an error. Also IMO if this happens k6 has a bug, as I am pretty sure we should stop sending data after we close the connection

Reply:
Good point, will look into this again. It seems we're sending a close twice, so this shouldn't be needed.

Reply:
So I'm starting to dislike gorilla/websocket...

Take a look at a8980133.

If not explicitly set, `websocket.Conn` [uses a custom close handler](https://github.com/loadimpact/k6/blob/05e8740b88325d15e853f6f37d26b2e9cf76e2e5/vendor/github.com/gorilla/websocket/conn.go#L1062-L1069) that is [called a few layers below `conn.ReadMessage()`](https://github.com/loadimpact/k6/blob/05e8740b88325d15e853f6f37d26b2e9cf76e2e5/vendor/github.com/gorilla/websocket/conn.go#L903).

We can't simply do `conn.SetCloseHandler(socket.closeConnection))` because that introduces race conditions when calling the Goja runtime between the parent `WS.Connect()` goroutine and the `go readPump(...)` (`conn.ReadMessage()`) one. I suppose we could add more synchronization here, but `Connect()` is already overly complicated, and I would instead suggest switching away from gorilla/websocket, for at least a couple of reasons:

- Its API is quite unfriendly, and as we've seen, it can be very easy to write non-compliant clients and servers.
- It doesn't make sense that a close handler is called while reading a message. This seems like a poorly thought-out feature since it forces the user to manually synchronize if specifying a custom close handler, as `ReadMessage()` should always be run in a separate goroutine.

At first glance it seems like [nhooyr/websocket](https://github.com/nhooyr/websocket) might be a good alternative with a saner API that would help us clean up the current implementation.

But let me know if you think there's a better way of avoiding the duplicate close issue than making `websocket.Conn.handleClose` a noop or doing this `err != websocket.ErrCloseSent` check.

Reply:
Given the fact that in the vast majority of cases (IMO) what gorilla is doing is what should be done and we should've read the docs to begin with, I would consider it not a great design, but not a poor one as well.
 I did take a look at the other lib and while I like it, given it comparisons below I am more inclined to go with the much more performant (apparently, I haven't benchmarked) https://github.com/gobwas/ws . But I agree with both you and the author that the gorilla one is very old and doesn't support a lot of things that it should, just because ... well it needs to be backwards compatible :) 

But in all cases ... I would consider this low prio and would probably want to do it when we have event loops and we are redoing the whole ws api, as I don't want to change some parts of the behaviour and break the current api :). 
Let's create an issue linking to this comments?
cc @na--  @cuonglm

Reply:
Agreed on the priority and leaving the refactor for later.

Let me know if I should create an issue about this.